e2: fix check of "pod is consistently pending"

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The E2E framework had no good support for checking that something remains in a certain state consistently. Several tests (ab)used "wait for pod running" with the expectation that it only fails because of a timeout. That's not a correct assumption and thus could have led to false negatives (test passes when some other error occurred that should have been treated as a test failure).

It also obscures the intent of the test.

A better way to test for "remains pending" is with gomega.Consistently and new infrastructure in the E2E framework that gets introduced in this PR.

#### Special notes for your reviewer:

This is a subset of a larger effort to revamp polling, see https://github.com/kubernetes/kubernetes/pull/113298. This PR can and should be merged first, regardless whether that other PR follows, because the changes make sense by themselves.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
e2e framework: add test case for early abort of pod polling

The early exit support was added in 004c1b644164601c99f51ce5a80b9fe9f5fd3534
but without a corresponding test case.

pohly committed:
e2e: enhance tests that check for pod staying pending

Using WaitTimeoutForPodRunningInNamespace followed by ExpectError was not very
precise (any error passed the check, not just the expected timeout) and
hard to read. Now the test's expectation is spelled out explicitly: the pod
must stay in pending.

pohly committed:
e2e framework: add NamespacedName

The E2E framework contains several functions which only differ in how they get
name and namespace: from an API object (WaitForPodRunningInNamespace) or as
separate parameters (WaitTimeoutForPodRunningInNamespace).

NamespacedName and the NamedObject interface enable writing helper functions
that can be called with both an API object (like *v1.Pod, which implements
Object and thus NamedObject) and name+namespace string (via
NamespacedName).

The other advantage of NamespacedName is that the order of name and namespace
parameter cannot be mixed up.

NamespacedName was derived from k8s.io/apimachinery/pkg/types.NamespacedName. A
separate type in the framework package was chosen a) to avoid additional
imports in test code and b) because the interface might not be suitable for
k8s.io/apimachinery.

pohly committed:
e2e framework: support polling with Get

These helper functions can be used in combination with
omega.Eventually/Consistently to implement polling of objects that is aware of
Kubernetes apiserver conventions:

- retry on certain errors instead of giving up,
  with "not found" handling decided by the caller (may or may not
  be fatal, depending on the test)
- sleep if requested by apiserver

pohly committed:
e2e framework/pod: add Get helper

This helper will be used by several different tests and makes those tests more
readable.

pohly committed:
e2e framework/pod: add gomega matchers

They can be used for polling with a get function and gomega.Eventually or
gomega.Consistently.

pohly commented:
This is not a complete fix. Other tests that use this pattern:
- PersistentVolumes-local Local volume that cannot be mounted [Slow] should fail due to non-existent path 
- PersistentVolumes-local Local volume that cannot be mounted [Slow] should fail due to wrong node

However, those use lots of additional helpers that all would need to be modified to separate "create pod" from "wait for it to be running". Let's address the current set of tests first, then deal with the rest in a separate PR.

pohly commented on test/e2e/common/storage/configmap_volume.go:
```diff
@@ -556,17 +556,19 @@ var _ = SIGDescribe("ConfigMap", func() {
-		framework.ExpectError(err, "created pod %q with non-optional configMap in namespace %q", pod.Name, f.Namespace.Name)
+		pod := createNonOptionalConfigMapPod(ctx, f, volumeMountPath)
+		getPod := e2epod.Get(f.ClientSet, pod)
+		gomega.Consistently(ctx, getPod).WithTimeout(framework.NewTimeoutContextWithDefaults().PodStart).Should(e2epod.BeInPhase(v1.PodPending))
```
PodStart = 5min is what was used before.

Note that https://github.com/kubernetes/kubernetes/pull/114783 will make it simpler to access that timeout.

pohly replied:
/hold

Let's use the timeouts from that PR.

pohly replied:
/hold cancel

This is done.

pohly commented:
/retest

sunnylovestiramisu commented:
> This is not a complete fix. Other tests that use this pattern:
> 
> * PersistentVolumes-local Local volume that cannot be mounted [Slow] should fail due to non-existent path
> * PersistentVolumes-local Local volume that cannot be mounted [Slow] should fail due to wrong node
> 
> However, those use lots of additional helpers that all would need to be modified to separate "create pod" from "wait for it to be running". Let's address the current set of tests first, then deal with the rest in a separate PR.

Should we create an issue to track this?

sunnylovestiramisu commented:
/lgtm

pohly commented:
> Should we create an issue to track this?

Not necessary, those tests fail in https://github.com/kubernetes/kubernetes/pull/113298 when always returning a nil error, so I have a reminder to fix them.

pohly commented:
https://storage.googleapis.com/k8s-triage/index.html?pr=1&text=pod%20pod-test%20found%20in%20namespace&test=Pods%20should%20run%20through%20the%20lifecycle%20of%20Pods%20and%20PodStatus is an unrelated flake.

/retest

aojea commented on test/e2e/framework/pod/wait.go:
```diff
@@ -830,18 +830,3 @@ func handleWaitingAPIError(err error, retryNotFound bool, taskFormat string, tas
 }
-
-// Decide whether to retry an API request. Optionally include a delay to retry after.
-func shouldRetry(err error) (retry bool, retryAfter time.Duration) {
```
I´ve already mentioned my concern when we added this , since client-go already implements this logic internally, adding 10 retries by default

pohly replied:
So you want to reopen the discussion around this?

I thought it had been settled when merging https://github.com/kubernetes/kubernetes/pull/109704 and https://github.com/kubernetes/kubernetes/commit/55ec0f2d9dd7a86dc1a487b035e725fd9d12721e.

We would have to pull in Tim again because I don't have a strong opinion on this and this particular logic is what he wanted.

pohly replied:
> we have a 2 loops now that retry x*10 on the same error

Just because we have two nested loops does not necessarily mean that there are more API requests. That depends on the delays that each loop is using. We don't control how quickly client-go retries.

What we decide here is whether we want to continue retrying for the entire duration of the overall timeout.

pohly replied:
https://github.com/kubernetes/kubernetes/pull/109816#issuecomment-1120892122

aojea replied:
:/ https://github.com/kubernetes/kubernetes/pull/111555

aojea replied:
> We don't control how quickly client-go retries.

every second https://github.com/kubernetes/kubernetes/pull/109816#issuecomment-1119332876

my point is that I don't think that a test that has to retry for 1 minute should pass

aojea replied:
but I'm not super strong just wanted to try another push XD

sunnylovestiramisu commented:
/lgtm

bart0sh commented:
/triage accepted
/priority important-longterm

SergeyKanzhelev commented:
/assign @haircommander

for node

haircommander commented on test/e2e/framework/get.go:
```diff
@@ -0,0 +1,131 @@
+	}
+}
+
+// GetObject takes a get function like clientset.CoreV1().Pods(ns).Get
```
this is a personal stylistic nit that I have that you're welcome to ignore:
I personally prefer when functions are defined above functions they call--so GetObject would be defined above HandleRetry. This is so the file can be read top to bottom and you see the call stack. Up to you if you wanna adopt it

pohly replied:
I usually do that, too, but I am not strict about it. When I believe that a file is easier to understand when introducing some concept first (like handling retries), then I might introduce that first and then build on it with other functions.

However, in this case I've been inconsistent. Either GetObject should come at the top (as you said) or at the end, after all the retry functions. I'll probably move it to the top.

haircommander commented:
couple of nits, none blocking, otherwise LGTM

pohly commented on test/e2e/framework/pod/wait.go:
```diff
@@ -168,6 +171,45 @@ func errorBadPodsStates(badPods []v1.Pod, desiredPods int, ns, desiredState stri
+				return true, nil
+			}
+		}),
+		BeRunning(),
```
In this case here, or in general?

I defined it to achieve feature parity with existing functions ("WaitFor..PodRunning"). It might save an import in some (rare) cases. Overall I agree that it doesn't add much value and could drop it.

haircommander replied:
yeah I mean generally

pohly replied:
Let's remove it. The explosion of `WaitFor*` functions with different names and subtle semantic differences has been a problem, we should strive to avoid that - thanks for reminding me! :grin:

pohly commented:
I updated this PR:
- addressed @haircommander's nits in https://github.com/kubernetes/kubernetes/compare/eb8ccf648217dfab854a0743793c7c0e2d6820d3..84d6d2c34145b771d6a4b0a39a39cd7e29bb3a60
- rebased onto latest master to get access to revised timeout support
- replaced `framework.NewTimeoutContextWithDefaults()` with `f.Timeouts`

@haircommander, @aojea: okay to merge?

pohly commented:
/hold

To give @aojea time for another look (not required), I'll lift the hold next week.

/assign @kubernetes/dep-approvers 

For approval of some new code being pulled from gomega. No version changes.

dims commented:
/approve
/lgtm

(changes in /vendor looks good to me)

aojea commented:
/hold cancel

I'm not able to review the gomega matchers since I don't have experience with that library, but I see other reviewers lgtm

Regarding gomega matches,  I personally don't think I'll be using them for the sig-network tests, but each SIG should be free to choose whatever they decide meanwhile they maintain their test healthy

Thanks for the fix and the great work Patrick
